### PR TITLE
Fix playback control issues with chromecast

### PR DESCRIPTION
**Pull request by Sky-High**
This PR solves various issues with medium playback with chromecast.

**Changes**
1. the navPlayingBar is not always shown and hidden properly at the begin and end of a track.
2. opening and immediately closing (cardbox) menu's without action yields an uncaught promise reject error
3. the |< rewind/previous track action does not always work properly
4. the previous issue is related to playlist management not being fully implemented for chromecast. This PR addresses some of the issues. Later PR's will fix more.
5. console.log calls in the chromecast plugin now more clearly refer to chromecast as origin module.

### CODE REVIEW

File: src/plugins/chromecastPlayer/plugin.js [hunk 1]
```diff
@@ -1079,6 +1109,15 @@ class ChromecastPlayer {
     getPlayerState() {
         return this.getPlayerStateInternal() || {};
     }
+
+    getCurrentPlaylistIndex() {
+        // tbd: update to support playlists and not only album with tracks
+        return this.getPlayerStateInternal()?.NowPlayingItem?.IndexNumber;
+    }
+
+    clearQueue(currentTime) { // eslint-disable-line no-unused-vars
```
Comment (thornbill) [inline src/plugins/chromecastPlayer/plugin.js:1118]: You should be able to just remove the parameter?

```suggestion
    clearQueue() {
```

File: src/components/remotecontrol/remotecontrol.js [hunk 2]
```diff
@@ -220,7 +220,10 @@ function updateNowPlayingInfo(context, state, serverId) {
                     itemContextMenu.show(Object.assign({
                         item: fullItem,
                         user: user
-                    }, options));
+                    }, options))
+                        .catch(function (err) {
+                            console.debug('[remotecontrol:updateNowPlayingInfo] click event reject', err);
+                        });
```
Comment (thornbill) [inline src/components/remotecontrol/remotecontrol.js:226]: same here

File: src/plugins/chromecastPlayer/plugin.js [hunk 3]
```diff
@@ -495,12 +500,17 @@ function getItemsForPlayback(apiClient, query) {
...
+ */
 function bindEventForRelay(instance, eventName) {
     Events.on(instance._castPlayer, eventName, function (e, data) {
-        console.debug('cc: ' + eventName);
-        const state = instance.getPlayerStateInternal(data);
-
-        Events.trigger(instance, eventName, [state]);
+        console.debug('[chromecastPlayer] ' + eventName);
+        // skip events without data
+        if (data && data.ItemId !== '' ) {
```
[the remark this answers is not in the document]
Comment (Sky-High) [reply src/plugins/chromecastPlayer/plugin.js:513]: check!

File: src/plugins/chromecastPlayer/plugin.js [hunk 4]
```diff
@@ -663,7 +687,7 @@ class ChromecastPlayer {
 
         normalizeImages(data);
 
-        console.debug(JSON.stringify(data));
+        // console.debug(JSON.stringify(data));
```
[the remark this answers is not in the document]
Comment (Sky-High) [reply src/plugins/chromecastPlayer/plugin.js:701]: Agree. This debug is called few times per second and clutters the log. So the log is clearer without the statement. and can be added again easily if specifically needed.

File: src/plugins/chromecastPlayer/plugin.js (hunk 1 above)
Comment (Sky-High) [reply src/plugins/chromecastPlayer/plugin.js:1129]: The current implementation of queue management is still rather limited in jellyfin chromecast, so this method is not really important. I intend to extend that in future PR's, and then the method is needed as playbackmanager expects it:
https://github.com/jellyfin/jellyfin-web/blob/4882d9c8cc931e38d422c4766bc0e5d351ca7f6a/src/components/playback/playbackmanager.js#L3968
I already added this method as placeholder for later, and playbackmanager seems to expect a parameter. So I prefer to leave this as is.

File: src/components/nowPlayingBar/nowPlayingBar.js [hunk 5]
```diff
@@ -170,17 +170,20 @@ function bindEvents(elem) {
 
     elem.querySelector('.previousTrackButton').addEventListener('click', function (e) {
         if (currentPlayer) {
-            if (lastPlayerState.NowPlayingItem.MediaType === 'Audio') {
+            if ( playbackManager.isPlayingAudio(currentPlayer) ) {
```
[the remark this answers is not in the document]
Comment (Sky-High) [reply src/components/nowPlayingBar/nowPlayingBar.js:173]: Agree

File: src/components/nowPlayingBar/nowPlayingBar.js [hunk 6]
```diff
@@ -574,7 +577,10 @@ function updateNowPlayingInfo(state) {
                         itemContextMenu.show(Object.assign({
                             item: item,
                             user: user
-                        }, options));
+                        }, options))
+                            .catch(function (err) {
+                                console.debug('[nowPlayingBar:updateNowPlayingInfo] click event reject', err);
+                            });
```
[the remark this answers is not in the document]
Comment (Sky-High) [reply src/components/nowPlayingBar/nowPlayingBar.js:583]: Fine with me!

File: src/components/remotecontrol/remotecontrol.js (hunk 2 above)
Comment (Sky-High) [reply src/components/remotecontrol/remotecontrol.js:226]: Id.

File: src/components/shortcuts.js [hunk 7]
```diff
@@ -121,14 +121,17 @@ function showContextMenu(card, options) {
...
+                    .then(result => {
+                        if (result.command === 'playallfromhere' || result.command === 'queueallfromhere') {
+                            executeAction(card, options.positionTo, result.command);
+                        } else if (result.updated || result.deleted) {
+                            notifyRefreshNeeded(card, options.itemsContainer);
+                        }
+                    })
+                    .catch(err => {
+                        console.debug('[shortcuts.showContextMenu] itemContextMenu.show reject', err);
+                    });
```
[the remark this answers is not in the document]
Comment (Sky-High) [reply src/components/shortcuts.js:134]: Check

File: src/components/remotecontrol/remotecontrol.js [hunk 8]
```diff
@@ -764,17 +767,20 @@ export default function () {
 
         context.querySelector('.btnPreviousTrack').addEventListener('click', function (e) {
             if (currentPlayer) {
-                if (lastPlayerState.NowPlayingItem.MediaType === 'Audio') {
+                if ( playbackManager.isPlayingAudio(currentPlayer) ) {
```
[the remark this answers is not in the document]
Comment (Sky-High) [reply src/components/remotecontrol/remotecontrol.js:776]: Ok.

File: src/controllers/itemDetails/index.js [hunk 9]
```diff
@@ -1966,13 +1966,17 @@ export default function (view, params) {
...
+                    .then(function (result) {
+                        if (result.deleted) {
+                            appRouter.goHome();
+                        } else if (result.updated) {
+                            reload(self, view, params);
+                        }
+                    })
+                    .catch(function(err) {
+                        console.debug('[index:onMoreCommandsClick] reject', err);
+                    });
```
[the remark this answers is not in the document]
Comment (Sky-High) [reply src/controllers/itemDetails/index.js:1979]: Id.

**Review by thornbill — APPROVED** [review]

**Pull request merged**